File: src/windows/Maple.WinUI/MapleUI/Organisms/MuiVersionHistoryPanel.cs
using System;
using System.Collections.Generic;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Maple.UI.Atoms;

namespace Maple.UI
{
    /// <summary>One saved version of the current asset's edit state.</summary>
    public sealed record MuiAssetVersion(string Id, string Label, DateTimeOffset SavedAt, bool IsCurrent = false);

    /// <summary>
    /// Maple.UI Version History Panel organism (unified-component-catalog.md
    /// §4.3, "Version History Panel" row: "Browse and restore versions",
    /// built from List Row, Timestamp, Button, Dialog) — one
    /// <see cref="MuiListRow"/> per saved version (trailing content is a
    /// <see cref="MuiTimestamp"/> plus a Restore button, current version
    /// marked and un-restorable), gated by a confirm
    /// <see cref="MuiDialog"/> before actually restoring.
    /// </summary>
    public sealed class MuiVersionHistoryPanel : ContentControl
    {
        public static readonly DependencyProperty VersionsProperty =
            DependencyProperty.Register(nameof(Versions), typeof(IReadOnlyList<MuiAssetVersion>), typeof(MuiVersionHistoryPanel),
                new PropertyMetadata(null, (d, _) => ((MuiVersionHistoryPanel)d).Rebuild()));

        public IReadOnlyList<MuiAssetVersion>? Versions
        {
            get => (IReadOnlyList<MuiAssetVersion>?)GetValue(VersionsProperty);
            set => SetValue(VersionsProperty, value);
        }

        public event EventHandler<string>? VersionRestored;

        private readonly StackPanel _root = new() { Orientation = Orientation.Vertical, Spacing = 2 };
        private readonly StackPanel _rows = new() { Orientation = Orientation.Vertical, Spacing = 2 };
        private readonly MuiEmptyState _empty = new() { IconName = "history", Title = "No earlier versions" };
        private readonly MuiDialog _confirmDialog = new() { Variant = MuiDialogVariant.Confirm, Title = "Restore this version?", Message = "Your current edits will be replaced.", ConfirmLabel = "Restore" };
        private string? _pendingRestoreId;

        public MuiVersionHistoryPanel()
        {
            _root.Children.Add(_rows);
            _root.Children.Add(_empty);
            _root.Children.Add(_confirmDialog);
            Content = _root;

            _confirmDialog.Confirmed += (_, _) =>
            {
                _confirmDialog.IsOpen = false;
                if (_pendingRestoreId is not null) VersionRestored?.Invoke(this, _pendingRestoreId);
            };
            _confirmDialog.Dismissed += (_, _) => _confirmDialog.IsOpen = false;

            Rebuild();
        }

        private void Rebuild()
        {
            var versions = Versions ?? Array.Empty<MuiAssetVersion>();
            _empty.Visibility = versions.Count == 0 ? Visibility.Visible : Visibility.Collapsed;
            _rows.Visibility = versions.Count > 0 ? Visibility.Visible : Visibility.Collapsed;

            _rows.Children.Clear();
            var now = DateTimeOffset.Now;
            foreach (var version in versions)
            {
                var trailing = new StackPanel { Orientation = Orientation.Horizontal, Spacing = 8 };
                trailing.Children.Add(new MuiTimestamp { Value = version.SavedAt, Now = now, Format = MuiTimestampFormat.Relative });
                if (!version.IsCurrent)
                {
                    var restore = new MuiButton { Variant = MuiButtonVariant.Ghost, ButtonSize = MuiButtonSize.Sm, Label = "Restore" };
                    var versionId = version.Id;
                    restore.Click += (_, _) => { _pendingRestoreId = versionId; _confirmDialog.IsOpen = true; };
                    trailing.Children.Add(restore);
                }
                else
                {
                    trailing.Children.Add(new MuiBadge { Variant = MuiBadgeVariant.Signal, Value = "Current" });
                }

                _rows.Children.Add(new MuiListRow { Label = version.Label, IconName = "history", TrailingContent = trailing, Active = version.IsCurrent });
            }
        }
    }
}

File: src/apple/Packages/MapleUI/Sources/MapleUI/Organisms/MuiVersionHistoryPanel.swift
// MuiVersionHistoryPanel.swift — Maple UI Organisms · Inspectors & panels
// (unified-component-catalog.md §4.3). Browse and restore versions, built
// from List Row, Timestamp, Button, Dialog. Mirrors Presets Panel's
// confirm-then-emit shape: `restored` only fires after the confirm
// dialog round-trips, never straight off the row press.

import SwiftUI

public struct MuiVersionItem: Identifiable, Sendable {
    public let id: String
    public let label: String
    public let timestampValue: Date
    public let current: Bool

    public init(id: String, label: String, timestampValue: Date, current: Bool = false) {
        self.id = id
        self.label = label
        self.timestampValue = timestampValue
        self.current = current
    }
}

public struct MuiVersionHistoryPanel: View {
    public let versions: [MuiVersionItem]
    public let loading: Bool
    public let restored: ((String) -> Void)?

    @State private var pendingRestoreId: String?

    public init(versions: [MuiVersionItem], loading: Bool = false, restored: ((String) -> Void)? = nil) {
        self.versions = versions
        self.loading = loading
        self.restored = restored
    }

    public var body: some View {
        ZStack {
            Group {
                if loading {
                    MuiSpinner(placement: .centered, label: "Loading versions")
                } else if versions.isEmpty {
                    MuiEmptyState(icon: "clock.arrow.circlepath", title: "No versions", message: "Every edit will show up here.")
                } else {
                    ScrollView {
                        LazyVStack(spacing: 0) {
                            ForEach(versions) { version in
                                MuiListRow(label: version.label, timestampValue: version.timestampValue, active: version.current, trailing: {
                                    if version.current {
                                        MuiBadge(variant: .signal, value: "Current")
                                    } else {
                                        MuiButton(label: "Restore", variant: .ghost, size: .sm) { pendingRestoreId = version.id }
                                    }
                                })
                            }
                        }
                    }
                }
            }

            MuiDialog(
                isPresented: pendingRestoreId != nil,
                title: "Restore this version?",
                message: "Your current edits will be replaced.",
                variant: .confirm,
                confirmLabel: "Restore",
                confirmed: { _ in confirmRestore() },
                dismissed: { pendingRestoreId = nil }
            )
        }
    }

    private func confirmRestore() {
        if let id = pendingRestoreId { restored?(id) }
        pendingRestoreId = nil
    }
}

#Preview("MuiVersionHistoryPanel") {
    MuiVersionHistoryPanel(versions: [
        MuiVersionItem(id: "1", label: "Current edit", timestampValue: Date(), current: true),
        MuiVersionItem(id: "2", label: "Before color grade", timestampValue: Date().addingTimeInterval(-3600)),
        MuiVersionItem(id: "3", label: "Original import", timestampValue: Date().addingTimeInterval(-86400)),
    ])
    .frame(width: 280, height: 220)
    .background(MuiTokens.bg)
}

#Preview("MuiVersionHistoryPanel — Loading / Empty") {
    VStack(spacing: 0) {
        MuiVersionHistoryPanel(versions: [], loading: true).frame(height: 100)
        MuiDivider()
        MuiVersionHistoryPanel(versions: []).frame(height: 140)
    }
    .background(MuiTokens.bg)
}
